File: training-data/cpp/gradient_descent.cpp
#include <functional>
#include <iostream>

double gradientDescent(std::function<double(double)> gradFn, double start, double learningRate, int iterations) {
    double x = start;
    for (int i = 0; i < iterations; i++) x -= learningRate * gradFn(x);
    return x;
}

int main() {
    double minimum = gradientDescent([](double x) { return 2 * (x - 3); }, 0, 0.1, 50);
    std::cout << minimum << std::endl;
    return 0;
}

File: training-data/rust/gradient_descent.rs
fn gradient_descent<F: Fn(f64) -> f64>(grad_fn: F, start: f64, learning_rate: f64, iterations: u32) -> f64 {
    let mut x = start;
    for _ in 0..iterations {
        x -= learning_rate * grad_fn(x);
    }
    x
}

fn main() {
    let minimum = gradient_descent(|x| 2.0 * (x - 3.0), 0.0, 0.1, 50);
    println!("{:.4}", minimum);
}
